src/effects/VST3/VST3Utils.cpp

```diff
@@ -238,3 +245,132 @@ bool VST3Utils::ParseAutomationParameterKey(const wxString& key, Steinberg::Vst:
    return false;
 
 }
+
+
+const std::vector<VST3Wrapper::ParameterInfo>* VST3Wrapper::getParameterInfos() const
+{
+   if (mParameterInfos.size() == 0)
+   {
+      if (mEditController == nullptr)
+         return nullptr;
+
+      for (int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+      {
+         Steinberg::Vst::ParameterInfo vstParameterInfo{ };
+         if (mEditController->getParameterInfo(i, vstParameterInfo) == Steinberg::kResultOk)
+         {
+            // Add all parameterinfos, regardless of whether they can be automated or not,
+            // because different flags of them will be checked by different methods.
```

Comment:
I think Vitaly mentioned that the set of parameters might really be very large?

If the filtering by parameter types is moved into `ForEachParameter`, then more repetition of code can be eliminated, and fewer parameters would need to be remembered here.

If you need to filter differently in one place, then VisitParameters might take another argument indicating how to filter the parameters, and its default value could select the usual case.

Reply:
I have seen that there are about 2000 parameters for the gain plugin which I am using in my tests - so yes, very large (and I can not fathom why a gain plugin needs to have 2k parameters... but that's another issue).

Scanning 2k of them each time is definitively overkill, so I'd implement what you suggest. However, I think that to make it worth, I'd need to have not just one vector<parameterinfo> as we have now, but a set of vector<parameterinfo> e.g. one for the params which are automatable, one for the readonly ones - otherwise if I had only one vector, we'd be back to square one i.e. I'd have to scan all 2000 items of it anyway.

Reply:
Note this from Steinberg's header file, ivsteditcontroller.h:

```
		kIsReadOnly		 = 1 << 1,	///< parameter cannot be changed from outside the plug-in (implies that kCanAutomate is NOT set)
...
		kIsHidden		 = 1 << 4,	///< parameter should be NOT displayed and cannot be changed from outside the plug-in 
									///< (implies that kCanAutomate is NOT set and kIsReadOnly is set) [SDK 3.7.0]

		kIsProgramChange = 1 << 15,	///< parameter is a program change (unitId gives info about associated unit 
									///< - see \ref vst3ProgramLists)
		kIsBypass		 = 1 << 16	///< special bypass parameter (only one allowed): plug-in can handle bypass
									///< (highly recommended to export a bypass parameter for effect plug-in)
```

So, check my logic please: I think we can always ignore the hidden parameters.  Because, see how `parameterInfo.flags` is used.

In the places where only the automatable parameters are examined -- that implies that the read-only are excluded.  In `SyncParameters`, all (and only) the read-onlies are excluded (explicitly).  Therefore, these visits exclude all the hidden.

In `FillParameters`, when building a plain UI -- we may examine some read-only parameters.  But, we exclude the hidden ones.

Therefore:  we always ignore the hidden.

Maybe you can easily verify with some temporary code, that most of these Amplify parameters are hidden?

Reply:
Apologies, it looks like the vst3 gain plugin I used for the tests is very unusual - it has 2049 parameters but only 2 of them are "real" and automatable, i.e. gain amount and on/off - and none of the 2049 params is hidden. 

 I tried with another gain plugin and it has only two parameters. So maybe we can leave things as they are?